File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/adapters/zlink/handlers/CreateGameHandler.java
package systems.zlink.samples.tictactoe.server.play.adapters.zlink.handlers;

import static systems.zlink.framework.ZLinkAwait.await;

import systems.zlink.contracts.core.RoutingId;
import org.springframework.beans.factory.ObjectProvider;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkRequest;
import systems.zlink.framework.spots.ZLinkSpotManager;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.configuration.SampleSettings;
import systems.zlink.samples.tictactoe.server.play.application.gamecreation.TicTacToeGameCreator;
import systems.zlink.samples.tictactoe.server.play.adapters.zlink.spots.TicTacToeGame;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameReq;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameRes;

@ZLinkHandlerGroup(SampleNames.PlayChannel)
public final class CreateGameHandler {
    private final ObjectProvider<ZLinkSpotManager> spots;
    private final SampleSettings settings;
    private final TicTacToeGameCreator gameCreator;

    public CreateGameHandler(
        ObjectProvider<ZLinkSpotManager> spots,
        SampleSettings settings,
        TicTacToeGameCreator gameCreator) {
        this.spots = spots;
        this.settings = settings;
        this.gameCreator = gameCreator;
    }

    @ZLinkRequest(packetName = "CreateGameReq")
    public CreateGameRes create(CreateGameReq request) {
        TicTacToeGameCreator.GameRoom room = gameCreator.nextRoom(request.gameName());
        await(spots.getObject().create(TicTacToeGame.class, RoutingId.from(room.roomId())));
        return new CreateGameRes(
            room.roomId(),
            settings.playEndpoint(),
            room.gameName());
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/adapters/zlink/handlers/CreateGameHandler.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.adapters.zlink.handlers

import kotlinx.coroutines.future.await
import org.springframework.beans.factory.ObjectProvider
import systems.zlink.contracts.core.RoutingId
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.handlers.ZLinkRequest
import systems.zlink.framework.spots.ZLinkSpotManager
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.play.application.gamecreation.TicTacToeGameCreator
import systems.zlink.samples.kotlin.tictactoe.server.play.adapters.zlink.spots.TicTacToeGame
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameRes

@ZLinkHandlerGroup(SampleNames.PlayChannel)
class CreateGameHandler(
    private val spots: ObjectProvider<ZLinkSpotManager>,
    private val settings: SampleSettings,
    private val gameCreator: TicTacToeGameCreator,
) {
    @ZLinkRequest
    suspend fun create(request: CreateGameReq): CreateGameRes {
        val room = gameCreator.nextRoom(request.gameName)
        spots.getObject().create(TicTacToeGame::class.java, RoutingId.from(room.roomId)).await()
        return CreateGameRes(
            roomId = room.roomId,
            gameName = room.gameName,
            ownerPlayEndpoint = room.ownerPlayEndpoint,
            playEndpoints = room.playEndpoints,
            playNodes = room.playNodes,
            requiredLevel = room.requiredLevel,
        )
    }
}
